### PR TITLE
Save failed replay And Fixed a bug

### PR DESCRIPTION
**Pull request by None1637**
Can save failed replay
Can save replay in offline
Fixed a bug that star calculate no calculate DT/NC/speed modify
move the pp display (score scene) to left-bottom

### CODE REVIEW

**Comment by Rian8337** [comment]
The star rating calculation bug has actually been fixed, but it's up to you if you want to revert the change related to it or not. It should be fine this way as well.

**Review by Rian8337 — CHANGES_REQUESTED** [review]

File: src/ru/nsu/ccfit/zuev/osu/helper/DifficultyReCalculator.java [hunk 1]
```diff
@@ -229,16 +229,20 @@ private boolean loadHitObjects(final BeatmapData data, float sliderSpeed, float
...
     }
 
-    public float recalculateStar(final TrackInfo track, float speedMultiplier, float cs) {
-        if (!init(track, speedMultiplier)) {
+    public float recalculateStar(final TrackInfo track, float cs){
+        return recalculateStar(track, cs, 1.0f);
+    }
+
+    public float recalculateStar(final TrackInfo track, float cs, float speed) {
+        if (!init(track)) {
```
[the remark this answers is not in the document]
Comment (None1637) [reply src/ru/nsu/ccfit/zuev/osu/helper/DifficultyReCalculator.java:237]: ok, I will fixed this problem

**Comment by Rian8337** [comment]
After testing in-game, I have two concerns:
- In replay, sliders occasionally gives 100 despite not being pressed. The video below should show it (bad audio, but the point is still there).
https://user-images.githubusercontent.com/52914632/111894058-3103c900-8a3a-11eb-8995-6d478fe31a14.mp4
- Every hitobject that appears/should appear after the play fails is counted as miss, which is not something that we want? Unless this is required to tackle possible replay bug.

**Review by Rian8337 — APPROVED** [review]

**Comment by Rian8337** [comment]
Just need answers for my questions.

**Comment by None1637** [comment]
> 
> 
> After testing in-game, I have two concerns:
> 
>     * In replay, sliders occasionally gives 100 despite not being pressed. The video below should show it (bad audio, but the point is still there).
>       https://user-images.githubusercontent.com/52914632/111894058-3103c900-8a3a-11eb-8995-6d478fe31a14.mp4
> 
>     * Every hitobject that appears/should appear after the play fails is counted as miss, which is not something that we want? Unless this is required to tackle possible replay bug.

In replay, sliders occasionally gives 100 despite not being pressed.
 -This is a bug. I think that's why replay can get higher combo than real play. Maybe I can fix this bug.
Every hitobject that appears/should appear after the play fails is counted as miss, which is not something that we want? Unless this is required to tackle possible replay bug.
 -If don't counted as miss, Every hitobject that appears/should appear after the play fails will be hit automatically in the replay. Count them as miss is the easiest way to solve this problem.

**Comment by Rian8337** [comment]
> This is a bug. I think that's why replay can get higher combo than real play. Maybe I can fix this bug.

Isn't that the original replay bug, which is caused by movement emulation?

> If don't counted as miss, Every hitobject that appears/should appear after the play fails will be hit automatically in the replay. Count them as miss is the easiest way to solve this problem.

The play would have failed already, so it shouldn't matter?

**Comment by None1637** [comment]
> Isn't that the original replay bug, which is caused by movement emulation? This one seems different since no cursor is present.

the original replay bug Include this bug.

> The play would have failed already, so it shouldn't matter? I reckon this had something to do with the bug we've discussed earlier.

It does have something to do with bugs, and If don't count as miss, the replay would not fail.

**Comment by NeroYuki** [comment]
i assume this can show off the bug better https://www.youtube.com/watch?v=0lfl53hofEQ (skip to 0:15)

**Comment by None1637** [comment]
> 
> 
> i assume this can show off the bug better https://www.youtube.com/watch?v=0lfl53hofEQ (skip to 0:15)

this bug is fixed in my previous [commit](https://github.com/osudroid/osu-droid/pull/102/commits/1914f4d60781834fe55500b2e93b44fe6c622434)

**Pull request merged**